src/renderer/components/+namespaces/namespace.store.ts

```diff
@@ -18,36 +18,111 @@ export const namespaceUrlParam = createPageParam<string[]>({
   }
 });
 
+export function getDummyNamespace(name: string) {
+  return new Namespace({
+    kind: Namespace.kind,
+    apiVersion: "v1",
+    metadata: {
+      name,
+      uid: "",
+      resourceVersion: "",
+      selfLink: `/api/v1/namespaces/${name}`
+    }
+  });
+}
+
 @autobind()
 export class NamespaceStore extends KubeObjectStore<Namespace> {
   api = namespacesApi;
-  contextNs = observable.array<string>();
+
+  @observable contextNs = observable.array<string>();
+  @observable isReady = false;
+
+  whenReady = when(() => this.isReady);
 
   constructor() {
     super();
     this.init();
   }
 
-  private init() {
-    this.setContext(this.initNamespaces);
-
-    return reaction(() => this.contextNs.toJS(), namespaces => {
-      storage.set(namespaces); // save to local-storage
-      namespaceUrlParam.set(namespaces, { replaceHistory: true }); // update url
-    }, {
-      fireImmediately: true,
+  onContextChange(callback: (contextNamespaces: string[]) => void, opts: IReactionOptions = {}): IReactionDisposer {
+    return reaction(() => this.contextNs.toJS(), callback, {
       equals: comparer.identity,
+      ...opts,
     });
   }
 
+  private async init() {
```

Comment:
Please consider how this function could be simlified.
1. It does a lot of mixed things - checks for hosted cluster, sets ready flag, introduces and creates disposers... It should be separated by couple of independent functions (I know they will be called only from here) to shrink cognitive load. This will allow also to remove comments. For example comment 
```// save selected namespaces to local-storage and update URL```
can be replaced by functions `saveNamespacesToStorage()`  and `updateURL()` etc.
2. Function with name `init()` shouldn't return anything? If it does, it should be called like `getNamespaceDisposers` or something like that?
3. Why we return disposers at the end if we don't use them anywhere?

Reply:
1 - maybe
2,3 - `return` currently does nothing, but imho it's always better when function return something to cancel some operation running from the call (disposer, event stop, etc.)